symbolic-ppdb/src/format/sourcelinks.rs

```diff
@@ -0,0 +1,186 @@
+        FormatError {
+            kind: FormatErrorKind::InvalidSourceLinkJson,
+            source: None,
+        }
+    }
+
+    /// Sort internal rules. This must be called before [Self::resolve].
+    fn sort(&mut self) {
+        // Put Exact matches first, then sort by the Prefix length, longest to shortest.
+        self.rules.sort_unstable_by(|a, b| match &a.pattern {
+            Pattern::Exact(_) => Ordering::Less,
+            Pattern::Prefix(a) => match &b.pattern {
+                Pattern::Exact(_) => Ordering::Greater,
+                Pattern::Prefix(b) => b.len().cmp(&a.len()),
+            },
+        });
+    }
+
+    /// Resolve the path to a URL.
+    pub fn resolve(&self, path: &str) -> Option<String> {
+        // Note: this is currently quite simple, just pick the first match. If we needed to improve
+        // performance in the future because we encounter PDBs with too many items, we can do a
+        // prefix binary search, for example.
+        let path_lower = path.to_lowercase();
+        for rule in &self.rules {
+            match &rule.pattern {
+                Pattern::Exact(value) => {
+                    if value == &path_lower {
+                        return Some(rule.url.clone());
+                    }
+                }
+                Pattern::Prefix(value) => {
+                    if path_lower.starts_with(value) {
+                        let replacement = path
+                            .get(value.len()..)
+                            .unwrap_or_default()
+                            .replace('\\', "/");
+                        return Some(rule.url.replace('*', &replacement));
+                    }
+                }
+            }
+        }
+        None
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_invalid_json() {
+        let mut mappings = SourceLinkMappings::empty();
+        assert!(mappings.add_mappings("").is_err());
+        assert!(mappings.add_mappings("foo").is_err());
+        assert!(mappings.add_mappings("{\"docs\": {\"k\": \"v\"}}").is_err());
+        assert!(mappings
+            .add_mappings("{\"documents\": [\"k\", \"v\"]}")
+            .is_err());
+        assert_eq!(mappings.rules.len(), 0);
+    }
+
+    #[test]
+    fn test_mapping() {
+        let mappings = SourceLinkMappings::new(
+            [r#"
+                {
+                    "documents": {
+                        "C:\\src\\*":                   "http://MyDefaultDomain.com/src/*",
+                        "C:\\src\\fOO\\*":              "http://MyFooDomain.com/src/*",
+                        "C:\\src\\foo\\specific.txt":   "http://MySpecificFoodDomain.com/src/specific.txt",
+                        "C:\\src\\bar\\*":              "http://MyBarDomain.com/src/*"
+                    }
+                }
+                "#, r#"
+                {
+                    "documents": {
+                        "C:\\src\\file.txt": "https://example.com/file.txt"
```

Comment:
can you also add an example with unix paths, just to be complete?

Reply:
👍